File: 0034-find-first-and-last-position-of-element-in-sorted-array/solution.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int n= nums.length;
        
        // finding 1st occurence
        int first = -1;
        int s=0, e = n-1;
        while( s<=e){
            int mid = (s+e)/2;

            if( nums[mid] == target){
                 first = mid;
                 e =mid-1;
            }else if( nums[mid] > target){
                e = mid-1;
            }
            else{
                s = mid+1;
            }
        }

        int ans[] = new int[2];
        if( first == -1){
            ans[0] = -1;
            ans[1] = -1;
             return ans;
        }

        ///  finding last occurence
        int second = -1;
         s=0;
         e = n-1;
        while( s<=e){
            int mid = (s+e)/2;

            if( nums[mid] == target){
                 second = mid;
                 s =mid+1;
            }else if( nums[mid] > target){
                e = mid-1;
            }
            else{
                s = mid+1;
            }
        }
        ans[0] = first;
        ans[1] = second;
        return ans;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/solution.cpp
class Solution {
    int Find1stPosition(vector<int> &nums, int target){
        int ans = -1;
        int s=0, e=nums.size()-1;

        while(s <= e){
            int mid = s + (e-s)/2;

            if(nums[mid] == target){
                // this could be the 1st or may be not so store it and look in left
                ans = mid;
                e = mid-1;
            }else if(nums[mid] < target) s = mid+1;
            else e = mid-1;
        }

        return ans;
    }

    int FindLastPosition(vector<int> &nums, int target){
        int ans = -1;
        int s=0, e=nums.size()-1;

        while(s <= e){
            int mid = s + (e-s)/2;

            if(nums[mid] == target){
                // this could be the last or may be not so store it and look in right
                ans = mid;
                s = mid+1;
            }else if(nums[mid] < target) s = mid+1;
            else e = mid-1;
        }

        return ans;
    }
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int i = Find1stPosition(nums, target);
        if(i == -1) return {-1, -1};
        int j = FindLastPosition(nums, target);
        return {i, j};
    }
};
